### PR TITLE
[multistage] initial commit to support all scalar/transform/filter match

### PR DESCRIPTION
**Pull request by walterddr**
goal is mainly for supporting functions post-aggregate and in v2 engine
- currently v2 engine doesn't support scalar function aliases (as it doesn't know to canonicalize function name during calcite operator loading)
- some transform/filter functions doesn't have matching scalar function signature and thus calcite can't parse them correctly with strong type info

This PR:
- address the signature issue by:
  - create placeholder function if those can be implemented in the future
  - ignore ones that don't make sense post transform
- rewrite some aliases list so that they can be understood by V2 engine as well.
- added tests on each of the above categories, to ensure feature e2e correctness (exhausted list of supported functions will be added later)

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java [hunk 1]
```diff
@@ -147,4 +157,48 @@ public static FunctionInfo getFunctionInfo(String functionName, int numParameter
...
+    public static long timeConvert(String timeConvertInput, String fromFormat, String toFormat) {
+      throw new UnsupportedOperationException("Placeholder scalar function, should not reach here");
+    }
+
+    public static long extract(String extractFormat, long timestamp) {
+      throw new UnsupportedOperationException("Placeholder scalar function, should not reach here");
+    }
+
+    // CHECKSTYLE:OFF
+    // @formatter:off
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java:189]: Why do we need to turn them on?

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java [hunk 2]
```diff
@@ -147,4 +157,48 @@ public static FunctionInfo getFunctionInfo(String functionName, int numParameter
...
+    public static boolean text_contains(String text, String pattern) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java:190]: Will calcite match the underscore? If so, other things can also break such as `json_extract_scalar`

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java [hunk 3]
```diff
@@ -25,7 +25,7 @@
 
 
 public enum TransformFunctionType {
-  // Aggregation functions for single-valued columns
+  // arithmetic functions for single-valued columns
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:28]: (format) Keep the comment capitalized to be consistent

**Review by Jackie-Jiang — COMMENTED** [review]
Please mark this as multistage and add some test on the multistage engine side

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java (hunk 2 above)
Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java:190]: +1. will fix this by adding annotation but only register to calcite map

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java (hunk 1 above)
Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java:189]: no need. have to fix alias anyway

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java [hunk 4]
```diff
@@ -79,6 +79,9 @@ private FunctionRegistry() {
         }
       }
     }
+    for (Method method : PlaceholderScalarFunctions.class.getMethods()) {
```
Comment (agavra) [inline pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java:82]: thanks for this PR! can you add some test cases to `QueryRunnerTest` to validate that functions registered in this way are picked up end-to-end?

Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java:82]: yup wilco

**Review by siddharthteotia — APPROVED** [review]

**Pull request merged**